File: REL.py
# import numpy as np
# from PIL import Image
# import matplotlib.pyplot as plt

# def threshold_binarization(image, threshold=90):
#     binary_image = np.where(image > threshold, 1, 0).astype(np.uint8)
#     return binary_image

# def merge_binary_data(binary_image, block_size=32):
#     height, width = binary_image.shape
#     merged_width = width // block_size
#     merged_data = np.zeros((height, merged_width), dtype=np.uint32)

#     for i in range(height):
#         for j in range(merged_width):
#             block = binary_image[i, j * block_size:(j + 1) * block_size]
#             merged_value = int(''.join(map(str, block)), 2)
#             merged_data[i, j] = merged_value

#     return merged_data

# def unmerge_binary_data(merged_data, block_size=32, original_width=320):
#     height, merged_width = merged_data.shape
#     binary_image = np.zeros((height, original_width), dtype=np.uint8)

#     for i in range(height):
#         for j in range(merged_width):
#             merged_value = merged_data[i, j]
#             binary_block = format(merged_value, f'0{block_size}b')
#             binary_image[i, j * block_size:(j + 1) * block_size] = list(map(int, binary_block))

#     return binary_image

# # 加载黑白图像
# original_image = Image.open("C:/Users/test/Desktop/RELTEST.jpg").convert("L")
# image_array = np.array(original_image)

# # 进行Otsu二值化
# # binary_image = otsu_binarization(image_array)
# # block_size = 64
# # compressed_data = merge_binary_data(binary_image, block_size)
# # decoded_image = unmerge_binary_data(compressed_data, block_size, original_width=image_array.shape[1])
# binary_image = threshold_binarization(image_array)

# compressed_data = merge_binary_data(binary_image)
# decoded_image = unmerge_binary_data(compressed_data)

# # 显示原始图像、Otsu's二值化后的图像和解压缩后的图像
# plt.figure(figsize=(16, 4))

# plt.subplot(1, 3, 1)
# plt.imshow(image_array, cmap='gray')
# plt.title('Original Image')

# plt.subplot(1, 3, 2)
# plt.imshow(binary_image, cmap='gray')
# plt.title("Otsu's Binarization")

# plt.subplot(1, 3, 3)
# plt.imshow(decoded_image, cmap='gray')
# plt.title('Decompressed Image')

# plt.show()



import numpy as np
from PIL import Image
import matplotlib.pyplot as plt

def sauvola_threshold(image, block_size, k, R):
    height, width = image.shape
    binary_image = np.zeros((height, width), dtype=np.uint8)

    for x in range(0, height, block_size):
        for y in range(0, width, block_size):
            block = image[x:x+block_size, y:y+block_size]
            block_mean = np.mean(block)
            block_std = np.std(block)
            threshold = block_mean * (1 + k * ((block_std / R) - 1))
            threshold_block = (block > threshold)
            binary_image[x:x+block_size, y:y+block_size] = threshold_block

    return binary_image

def merge_binary_data(binary_image, block_size=32):
    height, width = binary_image.shape
    merged_width = width // block_size
    merged_data = np.zeros((height, merged_width), dtype=np.uint32)

    for i in range(height):
        for j in range(merged_width):
            block = binary_image[i, j * block_size:(j + 1) * block_size]
            merged_value = int(''.join(map(str, block)), 2)
            merged_data[i, j] = merged_value

    return merged_data

def unmerge_binary_data(merged_data, block_size=32, original_width=320):
    height, merged_width = merged_data.shape
    binary_image = np.zeros((height, original_width), dtype=np.uint8)

    for i in range(height):
        for j in range(merged_width):
            merged_value = merged_data[i, j]
            binary_block = format(merged_value, f'0{block_size}b')
            binary_image[i, j * block_size:(j + 1) * block_size] = list(map(int, binary_block))

    return binary_image

def compute_average_brightness(image, block_size=2, array_size=160):
    
    # # 将图像转换为NumPy数组
    # image_array = np.array(image)

    # 初始化数组
    brightness_array = np.zeros((array_size, array_size), dtype=int)

    # 划分图像成块，计算单位亮度，填充数组
    for i in range(array_size):
        for j in range(array_size):
            block = image[i * block_size:(i + 1) * block_size, j * block_size:(j + 1) * block_size]
            average_brightness = int(np.mean(block))
            brightness_array[i, j] = average_brightness

    return brightness_array

# def threshold_binarization(image, threshold=90):
#     binary_image = np.where(image > threshold, 1, 0).astype(np.uint8)
#     return binary_image


# 加载黑白图像
original_image = Image.open("C:/Users/test/Desktop/10.jpg").convert("L")
image_array = np.array(original_image)



# Sauvola自适应二值化
block_size = 64  # 调整块大小
k = 0.1  # 调整增益因子
R = 128  # 调整参数R
brightness=compute_average_brightness(image_array)
binary_image = sauvola_threshold(brightness, 16, k, R)
# binary_image =threshold_binarization(brightness)
compressed_data = merge_binary_data(binary_image)
decoded_image = unmerge_binary_data(compressed_data)

# 显示原始图像、Sauvola二值化后的图像和解压缩后的图像
plt.figure(figsize=(16, 4))

plt.subplot(1, 3, 1)
plt.imshow(image_array, cmap='gray')
plt.title('Original Image')

plt.subplot(1, 3, 2)
plt.imshow(binary_image, cmap='gray')
plt.title("Sauvola Binarization")

plt.subplot(1, 3, 3)
plt.imshow(decoded_image, cmap='gray')
plt.title('Decompressed Image')

plt.show()

File: REL.c
#include <stdio.h>
#include <stdlib.h>

// Function to perform threshold binarization
void threshold_binarization(int** image, int** binary_image, int height, int width, int threshold) {
    for (int i = 0; i < height; i++) {
        for (int j = 0; j < width; j++) {
            binary_image[i][j] = (image[i][j] > threshold) ? 1 : 0;
        }
    }
}

// Function to merge binary data
void merge_binary_data(int** binary_image, int** merged_data, int height, int width, int block_size) {
    int merged_width = width / block_size;

    for (int i = 0; i < height; i++) {
        for (int j = 0; j < merged_width; j++) {
            int merged_value = 0;
            for (int k = 0; k < block_size; k++) {
                merged_value = (merged_value << 1) | binary_image[i][j * block_size + k];
            }
            merged_data[i][j] = merged_value;
        }
    }
}

// Function to unmerge binary data
void unmerge_binary_data(int** merged_data, int** binary_image, int height, int original_width, int block_size) {
    int merged_width = original_width / block_size;

    for (int i = 0; i < height; i++) {
        for (int j = 0; j < merged_width; j++) {
            int merged_value = merged_data[i][j];
            for (int k = 0; k < block_size; k++) {
                binary_image[i][j * block_size + k] = (merged_value >> (block_size - k - 1)) & 1;
            }
        }
    }
}

int main() {
    int height = 320;
    int width = 320;
    int threshold = 90;
    int block_size = 32;

    // Allocate memory for image and binary_image
    int** image = (int**)malloc(height * sizeof(int*));
    int** binary_image = (int**)malloc(height * sizeof(int*));

    for (int i = 0; i < height; i++) {
        image[i] = (int*)malloc(width * sizeof(int));
        binary_image[i] = (int*)malloc(width * sizeof(int));
    }

    // Fill image with data (you'll need to load your image data here)
    
    

    // Call threshold_binarization
    threshold_binarization(image, binary_image, height, width, threshold);

    // Call merge_binary_data
    int merged_width = width / block_size;
    int** merged_data = (int**)malloc(height * sizeof(int*));

    for (int i = 0; i < height; i++) {
        merged_data[i] = (int*)malloc(merged_width * sizeof(int));
    }

    merge_binary_data(binary_image, merged_data, height, width, block_size);

    // Call unmerge_binary_data to recover the binary_image

    // Free allocated memory

    return 0;
}
